rust/wal3/README.md

```diff
@@ -0,0 +1,468 @@
+wal3
+====
+
+wal3 is the write-ahead (lightweight) logging library.  It implements a linearlizable log that is
+built entirely on top of object storage.  It relies upon the atomicity of object storage to provide
+the If-Match header.  This allows us to create a log entirely on top of object storage without any
+other sources of locking or coordination.
+
+# Design
+
+wal3 is designed to work on object storage.  It is intended to to be lightweight, to allow a single
+machine to multiplex many logs simultaneously over a variety of paths.
+
+## Interface
+
+wal3 presents separate reader and writer interfaces in order to allow readers and writers to scale
+separately.  Readers can read the log without blocking writers and writers can append to the log
+without blocking readers.
+
+```text
+pub struct LogPosition {
+    // Offset is a sequence number indicating the total number of records inserted into the log.
+    pub offset: u64,
+    // Timestampl
+    pub timestamp; u64,
+}
+
+pub struct LogWriter { ... }
+
+impl LogWriter {
+    pub async fn open(options: LogWriterOptions) -> Result<Arc<Self>, Error>;
+    pub async fn append(self: &Arc<Self>, message: Message) -> Result<LogPosition, Error>;
+}
+
+// Limits allows encoding things like offset, timestamp, and byte size limits for the read.
+pub struct Limits { ... }
+
+pub struct LogReader { ...  }
+
+impl LogReader {
+    pub async fn open(options: LogReaderOptions) -> Result<Self, Error>;
+    pub async fn read(
+        self: &Self,
+        from: LogPosition,
+        limits: Limits,
+    ) -> Result<(LogPosition, Vec<Message>), Error>;
+}
+```
+
+The astute reader will note that this log is in process.  It is meant to be run under leader
```

Comment:
Curious, did we consider embedding it as a library in the query and compactor services? What trade-offs does that have?